src/webgpu/shader/validation/expression/call/builtin/refract.spec.ts

```diff
@@ -0,0 +1,158 @@
+const builtin = 'refract';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { Type, kConvertableToFloatVectors, scalarTypeOf } from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  ConstantOrOverrideValueChecker,
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatVectors);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() only errors in cases
+where a the calculations result in a non-representable value for the given type.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('a', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('b', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('c', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type = kValidArgumentTypes[t.params.type];
+    const scalarType = scalarTypeOf(kValidArgumentTypes[t.params.type]);
+    const vCheck = new ConstantOrOverrideValueChecker(t, scalarType);
+
+    // Refract equation:
+    //   let k = 1.0 - c * c * (1.0 - dot(b, a) * dot(b, a))
+    //   if (k < 0.0) { return vecN(0.0); }
+    //   return c * a - (c * dot(b, a) + sqrt(k)) * b
+    // Should be invalid if the calculations result in intermediate values that
+    // exceed the maximum representable float value for the given type.
+    const a = Number(t.params.a);
+    const b = Number(t.params.b);
+    const c = Number(t.params.c);
+
+    const b_dot_a = vCheck.checkedResult(b * a * type.width);
```

Comment:
Addition/subtraction operations are a potential source of may-fail behaviour, since OOB intermediates can be pull back in bounds by them.

Quickly looking at the examples here, the one_minus_b_dot_a_2, and c2_one_minus_b_dot_a_2, should be fine, since they are operating on non-negative values, due to the squaring, and the pull back in bounds can only occur when the operands have different signs (or are 0)... I guess you need to make sure one of the operands isn't 0 even for them.

For cdba_sqrt_k, I think there is a potential difference in signs, and thus potential may-fail. Specifically sqrt_k must be positive, but cdba can be negative, since I don't think the spec restricts the sign of c. I would take a closer look at that one to determine if it needs to be checked for may-fail.

Reply:
`checkedResult()` checks every value passed to it to see if it's in that ambiguous overflow zone, and skips the test internally if that's the case.

```js
  checkedResult(value: number): number {
    if (this.isAmbiguousOverflow(value)) {
      this.t.skip(`Checked value, ${value}, was within the ambiguous overflow rounding range.`);
    }

    const quantizedValue = this.quantize(value);
    if (!Number.isFinite(quantizedValue)) {
      this.#allChecksPassed = false;
    }
    return quantizedValue;
  }
```

Does that satisfy your concern here, or is there something more that I'm missing?

Reply:
Yeah it does